jmclient/jmclient/wallet.py

```diff
@@ -1013,6 +980,500 @@ def close(self):
+        if spent_outs is None:
+            # user has not provided input script information; psbt
+            # will not yet be usable for signing.
+            return new_psbt
+        for i, txinput in enumerate(new_psbt.inputs):
+            if spent_outs[i] is None:
+                # as above, will not be signable in this case
+                continue
+            if isinstance(spent_outs[i], (btc.CTransaction, btc.CTxOut)):
+                # note that we trust the caller to choose Tx vs TxOut as according
+                # to non-witness/witness:
+                txinput.utxo = spent_outs[i]
+            else:
+                assert False, "invalid spent output type passed into PSBT creator"
+        # we now insert redeemscripts where that is possible and necessary:
+        for i, txinput in enumerate(new_psbt.inputs):
+            if isinstance(txinput.utxo, btc.CTxOut):
+                # witness
+                if txinput.utxo.scriptPubKey.is_witness_scriptpubkey():
+                    # nothing needs inserting; the scriptSig is empty.
+                    continue
+                elif txinput.utxo.scriptPubKey.is_p2sh():
+                    try:
+                        path = self.script_to_path(txinput.utxo.scriptPubKey)
+                    except AssertionError:
+                        # this happens when an input is provided but it's not in
+                        # this wallet; in this case, we cannot set the redeem script.
+                        continue
+                    privkey, _ = self._get_priv_from_path(path)
+                    txinput.redeem_script = btc.pubkey_to_p2wpkh_script(
+                        btc.privkey_to_pubkey(privkey))
+        return new_psbt
+
+    def sign_psbt(self, in_psbt, with_sign_result=False):
+        """ Given a serialized PSBT in raw binary format,
+        iterate over the inputs and sign all that we can sign with this wallet.
+        NB IT IS UP TO CALLERS TO ENSURE THAT THEY ACTUALLY WANT TO SIGN
+        THIS TRANSACTION!
+        The above is important especially in coinjoin scenarios.
+        Return: (psbt, msg)
+        msg: error message or None
+        if not `with_sign_result`:
+        psbt: signed psbt in binary serialization, or None if error.
+        if `with_sign_result` True:
+        psbt: (PSBT_SignResult object, psbt (deserialized) object)
+        """
+        try:
+            new_psbt = btc.PartiallySignedTransaction.from_binary(in_psbt)
+        except Exception as e:
+            return None, "Unable to deserialize binary PSBT, error: " + repr(e)
+        privkeys = []
+        for k, v in self._utxos._utxo.items():
+            for k2, v2 in v.items():
+                privkeys.append(self._get_priv_from_path(v2[0]))
+        jmckeys = list(btc.JMCKey(x[0][:-1]) for x in privkeys)
+        new_keystore = btc.KeyStore.from_iterable(jmckeys,
+                                                  require_path_templates=False)
```

Comment:
If you are dealing with non-extended privkeys, `require_path_templates` is not relevant, no need to specify it. When you are dealing with extended keys, this also might not be a safe option, because this means that there will be no restrictions on the paths used to derive the keys -- the restrictions are set via path templates will be ignored. While these restrictions are more relevant to distinguishing change outputs, I think there might be the case where you want to sign with one path, but do not want to allow to sign with different path.

Reply:
I see. I will remove it next, thanks.

Reply:
Done in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/536/commits/a3bc5f7f03033e4fd6c8a9d2335a6f50aeccaf6e

Reply:
> the restrictions are set via path templates will be ignored

Sorry, misleading info. Not ignored, just `require_path_templates=False` will allow to use xkeys without also supplying path templates for them, so basically just the extra precautions will not be enforced.